lib/ssl/src/ssl.erl

```diff
@@ -1408,39 +1408,15 @@ clear_pem_cache() ->
     ssl_pem_cache:clear().
 
 %%---------------------------------------------------------------
--spec format_error({error, Reason}) -> string() when
+-spec format_error({error, Reason} | Reason) -> string() when
```

Comment:
I think the type is semantically correct but maybe too general?
Essentially, as a user who has never used this function, I would need to read the source code to see what I can put as input argument. I am not sure if this is the intended case. If not, I think the following type specs could "help", although they are a bit verbose:

```erlang
-spec format_error({error, ErrorReason} | Reason) -> string() when
      Reason :: string()
              | closed
              | {tls_alert, {_, Description :: string()}}
              | {options, Options :: term()}
              | {options, {socket_options, Option :: term()}}
              | {options, {socket_options, Option :: term(), Error}}
              | {options, {FileType, File :: string(), Error}}
              | InetError,
      FileType    :: cacertfile | certfile | keyfile | dhfile,
      Error       :: {error, ErrorReason} | ErrorReason,
      InetError   :: inet:posix()| system_limit,
      ErrorReason :: term().
```
and also:
```erlang
-spec do_format_error( string()
                     | closed
                     | {tls_alert, {_, Description :: string()}}
                     | {options, Options :: term()}
                     | {options, {socket_options, Option :: term()}}
                     | {options, {socket_options, Option :: term(), Error}}
                     | {options, {FileType, File :: string(), Error}}
                     | InetError) -> string()
              when
      FileType   :: cacertfile | certfile | keyfile | dhfile,
      Error      :: {error, ErrorReason} | ErrorReason,
      ErrorReason     :: term(),
      InetError :: inet:posix() | system_limit.
```

Reply:
Well, the purpose of this function is for the user to be able to provide a logging string from any error reason returned by any of the ssl API functions that may return {error, Reason}. For a legacy reason, the input can even be the tagged error tuple.  The whole point is that the user should not need to know what the error looks like but they want something that they can print in an error log. The reason may for instance include stack trace information for debugging purposes. The user will never need to create the input, it should be a matched Reason. We might want to spec the internal function for our own benefit, but I think that the documented value should be general.